Improve captions handling for streaming

friday opened the pull request:
Relates to #971

* Adds option to watch caption track changes and update language options
* Ensure languagechange and captionsenabled/captionsdisabled events are sent on initial setup, and that captionsenabled/captionsdisabled is only triggered on change.
* Ensure language is set in case the track is added after initialization
* Optimizations for code reuse.

`instance.toggleCaptions` will now ensure the elements are updated with the right classes/state even if they already were before. The reason for this is code reuse. I was able to remove the`captions.show` method this way. Without the changes to `instance.toggleCaptions` however it would not work if you load a new source in the same instance.

Another thing that may need explaining is `this.language` is being set in the new `captions.update` method (when it hasn't been set already). This is needed because setting `this.language` will rightfully bail if the language doesn't have a track (and log a warning which may need to be upgraded now). The track for the language may be added after initializing Plyr, so it has to try on every update until it has a match.

How to test:
```js
new Plyr(video, {captions: {update: true}});
video.addTextTrack('subtitles', 'Volapük', 'vo');
```

Hls.js introduces another dimension of issues to the table, and it's not really ours to fix, but if you want to test it, here's a new simplified jsbin (using this branch) with Plyr + hls.js and two kinds of captions:
http://jsbin.com/mayikihosu/edit?js,output

Not in this PR:
* It doesn't fix the css issue (I think?) causing #980, however it does ensure `player.language` isn't something that doesn't have a track. So the last bullet would be `undefined` not `es`
* Keeping an internal reference of the current track index - This would be nice for other streaming libraries as well since it seems to be how they operate (the same applies to quality options). The method I'm running in the `languagechange` listener in the jsfiddle examples is not safe without this, if there are duplicate tracks per language (like hard-of-hearing subtitles).

sampotts commented on readme.md:
```diff
@@ -303,7 +303,7 @@ Note the single quotes encapsulating the JSON and double quotes on the object ke
 | `toggleInvert`       | Boolean                    | `true`                                                                                                                         | Allow users to click to toggle the above.                                                                                                                                                                                ...
 | `listeners`          | Object                     | `null`                                                                                                                         | Allows binding of event listeners to the controls before the default handlers. See the `defaults.js` for available listeners. If your handler prevents default on the event (`event.preventDefault()`), the default hand ...
-| `captions`           | Object                     | `{ active: false, language: window.navigator.language.split('-')[0] }`                                                         | `active`: Toggles if captions should be active by default. `language`: Sets the default language to load (if available).                                                                                                 ...
+| `captions`           | Object                     | `{ active: false, language: browser language, update: false  }`                                                         | `active`: Toggles if captions should be active by default. `language`: Sets the default language to load (if available). `update`: Listen to changes to tracks and update menu. This is needed for some streaming libraries, bu ...
```
How do you feel about perhaps making the default language `'auto'` and we translate that into `navigator.language` in the code? Might be more user friendly.

friday replied:
I have actually already done that :), but that was for the SSR issue, and I ended up pushing a different solution. I agree this is better.

sampotts commented on src/js/ui.js:
```diff
@@ -55,8 +55,10 @@ const ui = {
 
-        // Captions
-        captions.setup.call(this);
+        // Setup captions for html5
```
Super minor but can we make this `HTML5`

friday commented:
I changed some CORS settings for our test server so https://testing.studiotogo.com/project/59db9312345a8234c27c809b/playlist.m3u8 can be used to test multiple languages and VTT (the other one is embedded).

(you can replace the url in the jsfiddle examples).

friday commented:
I changed this to WIP since I found another issue (`languagechange` event occasionally triggering before event listener).

I also updated the description to replace the hls.js stuff with a simple test. Hls.js introduces another dimension of issues to the table, and it's not really ours to fix.

sampotts commented:
Is this one ok to merge now? Awesome work btw 👍

friday commented:
Yes. I'm no longer working on it. I had to do some changes and more testing, but it seems solid now :)
